File: src/com/klloyds/java/Variance.java
package com.klloyds.java;

import com.sun.org.apache.xerces.internal.xs.datatypes.ObjectList;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

public class Variance {

    public static void main(String[] args) {

        // Covariance - Subtype ok
        final Animal animal = new Dog(); // ok
        // final kotlin.com.klloyds.kotlin.Dog dog = new kotlin.com.klloyds.kotlin.Animal(); // not ok
        // final kotlin.com.klloyds.kotlin.Dog dog = (kotlin.com.klloyds.kotlin.Dog) new kotlin.com.klloyds.kotlin.Animal(); // not ok - ClassCastException
        // kotlin.com.klloyds.kotlin.printName(animal); // ok


        // Java - Use-site com.klloyds.java.Variance

        // Covariance - Generics with wildcard "extends"
        final List<Integer> integers = new ArrayList<>();
        final List<? extends Number> numbers = integers;
        final Number number = numbers.get(0); // We can read from a generic structure
        //numbers.add(new Integer(1)); // not ok


        // Contravariance - Generics with wildcard "super"
        final List<Number> numbers1 = new ArrayList<>();
        List<? super Integer> integers1 = numbers1;
        integers1.add(new Integer(1)); // We can write to a generic structure
        //final Number number1 = integers1.get(0); // not ok

    }

    private static class Animal {
    }

    private static class Dog extends Animal {
    }

    private static void printName(Animal animal) {
        System.out.println(animal.getClass().getSimpleName());
    }
}

File: src/com/klloyds/kotlin/Variance.kt
package com.klloyds.kotlin

import java.util.ArrayList

fun main(args: Array<String>) {

    // Covariance - Subtype ok
    val animal: Animal = Dog() // ok
    // val dog: com.klloyds.kotlin.Dog = com.klloyds.kotlin.Animal() // not ok
    // val dog = com.klloyds.kotlin.Animal() as com.klloyds.kotlin.Dog // not ok - ClassCastException :(
    // com.klloyds.kotlin.printName(animal) // ok


    // Kotlin - Declaration-site com.klloyds.java.Variance

    // Covariance - Generics with "out" keyword / same as "extends"
    val integers: ArrayList<Int> = arrayListOf(1, 2, 3)
    val numbers: ArrayList<out Number> = integers
    val number: Number = numbers[0] // We can read from a generic structure
    //numbers.add(1) // not ok


    // Contravariance - Generics with "in" keyword / same as "super"
    val numbers1: ArrayList<Number> = arrayListOf(1, 2, 3)
    val integers1: ArrayList<in Int> = numbers1
    integers1.add(0, 4) // We can write to a generic structure
    //val number1: Number = integers1[0] // not ok

    // Invariance

}

open class Animal
class Dog : Animal()

fun printName(animal: Animal) {
    println(animal.javaClass.simpleName)
}
